### DISCUSSION TITLE
Using Arduino Audio Tools header files across multiple compilation units in PlatformIO

### DISCUSSION THREAD

**Discussion by nullpainter** [pschatzmann/arduino-audio-tools#1831] [Q&A] [answered]
Caveat: I am a C# developer and am still a bit fuzzy with C++, so apologies for any misunderstandings, incorrect terminology, etc.

I am using Arduino Audio Tools to process an I2S stream for FFT-based visualisation on a RGB matrix. I am using PlatformIO and have a reference to an `AudioRealFFT` object in a constructor for one of my classes.

`FFTDisplay.h` is included in both my `main.cpp` and in a header file for a class containing my visualisation code. As `FFTDisplay.h` contains not only the class definition but also implementations, I'm receiving a multiple definition error from the linker for the `selfFFTDisplay` pointer initialized in this header.

Short of copying this header file locally and splitting it out into a header + cpp file, is there a recommended approach to manage this?

**Comment by pschatzmann** [comment, marked as answer]
I [committed a correction](https://github.com/pschatzmann/arduino-audio-tools/commit/d60b19878b18878b7120dc5ff3839b5d86660cce)

**Comment by nullpainter** [comment]
Amazing, thank you!
